### ISSUE TITLE
salt-ssh imports unrelated python modules on both slave (fatal) and master

### ISSUE THREAD

**Issue by pfalcon** [saltstack/salt#7913]
~~~~
$ salt-ssh -c . localhost test.ping
localhost:
    ERROR: Permissions problem, target user may need to be root or use sudo:
     [CRITICAL] Unable to import msgpack or msgpack_pure python modules
    WARNING: '/etc/salt' directory does not exist.
    Failed to create directory path "/etc/salt/pki/minion" - [Errno 13] Permission denied: '/etc/salt'
~~~~

msgpack or msgpack_pure should not be needed for salt-ssh.

Also, running with -lall:
~~~~
Traceback (most recent call last):
  File "/home/pfalcon/projects-3rdparty/salt/.venv/local/lib/python2.7/site-packages/salt/loader.py", line 641, in gen_functions
    ), fn_, path, desc
  File "/home/pfalcon/projects-3rdparty/salt/.venv/local/lib/python2.7/site-packages/salt/log/handlers/logstash_mod.py", line 114, in <module>
    import zmq
  File "/usr/local/lib/python2.7/dist-packages/pyzmq-2.2.0.1-py2.7-linux-i686.egg/zmq/__init__.py", line 51, in <module>
    from zmq import core, devices
  File "/usr/local/lib/python2.7/dist-packages/pyzmq-2.2.0.1-py2.7-linux-i686.egg/zmq/core/__init__.py", line 26, in <module>
    from zmq.core import (constants, error, message, context,
ImportError: libzmq.so.3: cannot open shared object file: No such file or directory
~~~~

I have msgpack installed in virtualenv, so in this case master doesn't error out, but otherwise it still does - but there shouldn't be a need for msgpack or zmq with salt-ssh.

**Comment by jcftang** [comment]
I'm also experiencing this issue with salt-ssh 

<pre>
# salt-ssh 'node0100' cmd.run uptime
node0100:
    ----------
    retcode:
        1
    stderr:
        [CRITICAL] Unable to import msgpack or msgpack_pure python modules
        Traceback (most recent call last):
          File "/tmp/.root_6cb53b__salt/salt-call", line 4, in <module>
            salt_call()
          File "/tmp/.root_6cb53b__salt/salt/scripts.py", line 122, in salt_call
            client.run()
          File "/tmp/.root_6cb53b__salt/salt/cli/__init__.py", line 419, in run
            caller.run()
          File "/tmp/.root_6cb53b__salt/salt/cli/caller.py", line 225, in run
            ret = self.call()
          File "/tmp/.root_6cb53b__salt/salt/cli/caller.py", line 129, in call
            ret['return'] = func(*args, **kwargs)
          File "/tmp/.root_6cb53b__salt/salt/modules/cmdmod.py", line 623, in run
            fn_.write(serial.dumps(jid_dict))
          File "/tmp/.root_6cb53b__salt/salt/payload.py", line 115, in dumps
            return msgpack.dumps(msg)
        NameError: global name 'msgpack' is not defined
    stdout:
  </pre>

With this ...

<pre>
d# salt --versions-report
           Salt: 2014.7.0
         Python: 2.7.3 (default, Dec 18 2014, 19:10:20)
         Jinja2: 2.7.3
       M2Crypto: 0.21.1
 msgpack-python: 0.3.0
   msgpack-pure: 0.1.3
       pycrypto: 2.4.1
        libnacl: Not Installed
         PyYAML: 3.10
          ioflo: Not Installed
          PyZMQ: 13.0.0
           RAET: Not Installed
            ZMQ: 3.2.2
           Mako: Not Installed

</pre>

**Comment by basepi** [comment]
Thanks for the reports. I'm going to re-open this and find some time to track down these msgpack imports.

**Comment by jcftang** [comment]
@basepi I think I figured out what my problem is with salt-ssh for my case... Here's an example of the problem when I try running salt-call on the target

<pre>
/tmp/.root_6cb53b__salt# python
Python 2.7.3 (default, Mar 14 2014, 17:55:54) 
[GCC 4.6.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>>  msgpack.loads(msgpack.dumps([1, 2, 3]), use_list=True)
  File "<stdin>", line 1
    msgpack.loads(msgpack.dumps([1, 2, 3]), use_list=True)
    ^
IndentationError: unexpected indent
>>> import msgpack
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "msgpack/__init__.py", line 3, in <module>
    from msgpack._msgpack import *
ImportError: msgpack/_msgpack.so: wrong ELF class: ELFCLASS64
</pre>

Basically I am launching salt-ssh from my intel x86_64 based machine and salt-ssh is copying out some intel compiled objects to my target which is actually a debian armhf based machine. I guess the correct solution is check for the architecture of the target if its the same as the host running salt-ssh and run a python only version of msgpack if it's not the same architecture.

This problem probably affects hosts that are intending on running salt-ssh from a 64bit master  targetting a 32bit minion.

**Comment by basepi** [comment]
`msgpack` should actually not be required at all for salt-ssh. I think we're still including it in the thin just for safety, but what we really need to do is track down any ungated calls to msgpack and gate them for salt-ssh.

@jcftang @bbinet Can anyone reproduce any msgpack errors with the patch in #19742? It should be safe to remove msgpack completely from the thin, but I want to see if we can catch any bad imports of msgpack before we merge that.

**Comment by jcftang** [comment]
@basepi I can confirm that #19742 resolves the above for me, I did have to clear out the thin cached tarball and also /tmp on the target.

<pre>
~# salt-ssh 'node0100' cmd.run uptime
node0100:
     12:06:58 up 156 days,  2:02,  1 user,  load average: 0.08, 0.05, 0.05
</pre>

It at least works for cmd.run in the above simple test.

**Comment by basepi** [comment]
Thanks for the update, @jcftang. I'm going to close this again for now. Please comment if you see any msgpack errors in salt-ssh going forward, they should be quick and easy to fix if we've missed any.

**Comment by mventimi** [comment]
I'm seeing the same error when I bootstrap from git develop:
```
$ salt-ssh -i lab cmd.run 'uname -a'
lab:
    ----------
    retcode:
        1
    stderr:
        [CRITICAL] Unable to import msgpack or msgpack_pure python modules
        Traceback (most recent call last):
          File "/tmp/.pi_6787e8__salt/salt-call", line 4, in <module>
            salt_call()
          File "/tmp/.pi_6787e8__salt/salt/scripts.py", line 220, in salt_call
            client.run()
          File "/tmp/.pi_6787e8__salt/salt/cli/call.py", line 50, in run
            caller.run()
          File "/tmp/.pi_6787e8__salt/salt/cli/caller.py", line 235, in run
            ret = self.call()
          File "/tmp/.pi_6787e8__salt/salt/cli/caller.py", line 139, in call
            ret['return'] = func(*args, **kwargs)
          File "/tmp/.pi_6787e8__salt/salt/modules/cmdmod.py", line 617, in run
            jid_dict = serial.load(fn_)
          File "/tmp/.pi_6787e8__salt/salt/payload.py", line 112, in load
            return self.loads(data)
          File "/tmp/.pi_6787e8__salt/salt/payload.py", line 97, in loads
            return msgpack.loads(msg, use_list=True)
        NameError: global name 'msgpack' is not defined
    stdout:

$ salt --versions-report
           Salt: 2015.2.0-307-g1fc9a52
         Python: 2.6.9 (unknown, Sep 13 2014, 00:25:11)
         Jinja2: 2.7.2
       M2Crypto: 0.20.2
 msgpack-python: 0.4.5
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.11
          ioflo: Not Installed
          PyZMQ: 14.3.1
           RAET: Not Installed
            ZMQ: 3.2.4
           Mako: Not Installed
```
Could this be a regression? It doesn't happen with 2014.7.1:
```
$ salt-ssh -i lab cmd.run 'uname -a'
lab:
    Linux raspberrypi 3.12.28+ #709 PREEMPT Mon Sep 8 15:28:00 BST 2014 armv6l GNU/Linux

$ salt --versions-report
           Salt: 2014.7.1
         Python: 2.6.9 (unknown, Sep 13 2014, 00:25:11)
         Jinja2: 2.7.2
       M2Crypto: 0.20.2
 msgpack-python: 0.4.5
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.11
          ioflo: Not Installed
          PyZMQ: 14.3.1
           RAET: Not Installed
            ZMQ: 3.2.4
           Mako: Not Installed
```

**Comment by basepi** [comment]
@mventimi Thanks for the update. Would you mind opening a new issue for this one? Just include all the information you've included here.

**Comment by basepi** [comment]
@mventimi Yes, I've done a little bit of investigation into this and this is related to some changes made in `develop`. More to come when I get a chance to talk to @thatch45 next week, but we should definitely open a new issue.

**Comment by mventimi** [comment]
Thanks @basepi. Issue 20276 opened.
